parts/k8s/containeraddons/kubernetesmasteraddons-cluster-autoscaler-deployment.yaml

```diff
@@ -129,7 +129,7 @@ metadata:
   labels:
     app: cluster-autoscaler
     kubernetes.io/cluster-service: "true"
-    addonmanager.kubernetes.io/mode: EnsureExists
+    addonmanager.kubernetes.io/mode: Reconcile
```

Comment:
@feiskyer any reasons that you know of why we would NOT want to do this?

Reply:
With EnsureExists, users could update the config options later, e.g. change the node pools and scan intervals.

Reply:
Users will need to change the spec on the disk, ie. in /etc/kubernetes/addons in order to change the config. If this is set to EnsureExists the spec won't get updated during upgrades unless the deployment is deleted.

Reply:
To speak to @feiskyer's point:

It's true that the current cluster-autoscaler spec/implementation is artificially limited to "VMSS" + "one VMSS pool" (the first pool, to be specific). So, to use a concrete example, if I build a VMSS AKS Engine cluster with 3 pools, and I enable the cluster-autoscaler addon, I will have to (after the cluster is deployed) manually re-configure my cluster-autoscaler Deployment spec to add the add'l 2 node pools. So, changing this to `Reconcile` means that when the user upgrades his/her cluster that custom configuration will be overridden, and effectively those other 2 pools won't be enabled for cluster-autoscaler.

However, I think the root cause of this issue is the artificial limitation of the current cluster-autoscaler addon. It's not too much work to make it generally useful for all VMSS pools in the api model.

So, my proposal is that we move this to Reconcile to address the issue of the autoscaler implementation (especially the deployment image) becoming unexpectedly stale across cluster upgrades. And also, to add validation to ensure that customers with > 1 VMSS node pool aren't allowed to use this addon. In practice, there isn't much value here for those folks anyways compared to following the online docs (IMO).

And then medium term we actually make this spec useful for clusters with arbitrary numbers of node pools.

@CecileRobertMichon @mboersma thoughts?

Reply:
@jackfrancis I think cluster-autoscaler shouldn't be an addon in the current state, in our cluster we disabled the addon and used helm to install the cluster-autoscaler as we need always to add/modify some custom configs (e.g. scale down delay interval). if we use the addon then we need will every time to change the addon file on the master node which is not a good way of managing the configuration changes.
 
I would say addons should be only for components which rarely need configuration changes from the user, like azure npm, cni, etc..

Reply:
@mo-saeed I agree 100%. I do, however, think there is *some* convenience value in automatically enabling various componentry (e.g., cluster-autoscaler) with a simple api model flag. Which componentry we choose to enable as addons (which as you point out does not expose any production-useful configuration interfaces like helm) should really be up to our users to decide. (especially users who aren't sophisticated enough to curate their own helm charts/repositories)

But yeah, if more folks tell us that they don't use this addon then I agree we should move our investments along this vector elsewhere.

Reply:
It shouldn't be too hard to evolve the addon to allow for more than one scale set (and also make that configurable potentially). 

Agreed with the short term fix plan of setting to `Reconcile` (hence the PR :) ) . However, I'm not sure we need to add validation to disallow more that one scale set as it might be useful to keep that as a possibility (for testing or for quickly deploying the autoscaler and then changing its config on the cluster) in the meantime. Our docs do mention that this addon shouldn't be used with more than one agent pool. https://github.com/Azure/aks-engine/blob/master/examples/addons/cluster-autoscaler/README.md#L16

Reply:
btw in case you don't already know this @mo-saeed (you probably do), the cluster-autoscaler addon can be configured to change scan-interval, etc.:

```
"addons": [
          {
            "name": "cluster-autoscaler",
            "enabled": true,
            "config": {
              "min-nodes": "1",
              "max-nodes": "5",
              "scan-interval": "10s"
            }
          }
        ]
```

Reply:
@CecileRobertMichon Thank you ! that was just an example and it's nice to have this configurable but my point is configuration management after first installation, which i would love some day to be supported by aks-engine command (sth like `aks-engine update`. for example changing api-server, kubelet or even addons flags/config so i don't need to go the master nodes and change files manually which can lead sometimes to problems.
I also understand @jackfrancis and your point for the value of having it as addon for some users :)

Reply:
Definitely, agree 💯 %